File: src/double_precision/javascript/qh_1.js
Qh=function(f,a,b,h=1/64){
  var exp=Math.exp ;
  var E=function(x){return(exp(-exp(x*(1+exp(-x))))) ;} ;
  var Ep=function(x){return(exp(-exp(x*(1+exp(-x))))* exp(x*(1+exp(-x)))* (1+(1-x)*exp(-x))) ;} ;
  if(a==b){return(0) ; } 
  if(a>b){return(-Qh(f,b,a,h)); }
  if(a==-Infinity && b==Infinity){
    return(Qh(x=> f(x/(1-x**2))*(1+x**2)/(1-x**2)**2,-1,1,h)) ;
  }
  if(isFinite(a) && b==Infinity){
    return(Qh(x=> f(a-1+1/x)/x**2,0,1,h)) ;
  }
  if(a==-Infinity && isFinite(b)){
    return(Qh(x=> f(b+1-1/x)/x**2,0,1,h)) ;
  }
  var ba=b-a ;
  var id=function(x){return(isFinite(x) ? x : 0) ;} ;
  var g=function(x){return(ba*id(f(a+ba*E(x))*Ep(x))) ;} ;
  var S=h*g(0) ;
  var n=-1 ; var x=-h ; var SS=S+h*g(x) ;
  while(SS!=S || n>-10){
    n-- ; x -= h ; S=SS ; SS += h*g(x) ;
  }
  var N=-n ; n=1 ; x=h ; SS=S+h*g(x) ;
  while(SS!=S || n<10){
    n++ ; x += h ; S=SS ; SS += h*g(x) ;
  }
  N += n+1 ;
  console.log(N) ;
  return(S) ;
}

File: src/double_precision/c_language/qh_1.c
#include <stdio.h>
#include <math.h>

double Qh01(double f(double),double h){
  double x, Ex, Epx, fx, S, SS;
  int n;
  x=0;
  Ex=exp(-exp(x*(1+exp(-x))));
  Epx=Ex*exp(x*(1+exp(-x)))*(1+(1-x)*exp(-x));
  S=h*f(Ex)*Epx;
  n=-1; x=-h;
  Ex=exp(-exp(x*(1+exp(-x))));
  Epx=Ex*exp(x*(1+exp(-x)))*(1+(1-x)*exp(-x));
  fx=f(Ex)*Epx;
  SS=S+h*fx;
  while((SS!=S)||(n>-10)){
    n-=1; x-=h;
    S=SS;
    Ex=exp(-exp(x*(1+exp(-x))));
    Epx=Ex*exp(x*(1+exp(-x)))*(1+(1-x)*exp(-x));
    fx=f(Ex)*Epx;
    SS +=h*fx;
  }
  n=1; x=h; 
  Ex=exp(-exp(x*(1+exp(-x))));
  Epx=Ex*exp(x*(1+exp(-x)))*(1+(1-x)*exp(-x));
  fx=f(Ex)*Epx;
  SS=S+h*fx;
  while((SS!=S)||(n<10)){
    n++; x+=h;
    S=SS;
    Ex=exp(-exp(x*(1+exp(-x))));
    Epx=Ex*exp(x*(1+exp(-x)))*(1+(1-x)*exp(-x));
    fx=f(Ex)*Epx;
    SS +=h*fx;
  }
  return S;
}

double Qhinfinf(double f(double),double h){
  double g(double x){
    double x1mx=x*(1-x);
    double x1mx2=x1mx*x1mx;
    return f((x-0.5)/x1mx)*(0.5-x1mx)/x1mx2;
  }
  return Qh01(g,h);
}

double Qhainf(double f(double), double a, double h){
  double g(double x){
    return f(a-1+1/x)/x/x;
  }
  return Qh01(g,h);
}

double Qhinfb(double f(double), double b, double h){
  double g(double x){
    return f(b+1-1/x)/x/x;
  }
  return Qh01(g,h);
}

double f1(double x){
  return 1;
}

double f2(double x){
  return x;
}

double f3(double x){
  return pow(x,99);
}

double f4(double x){
  return exp(x);
}

double f5(double x){
  return 1/(1+x*x) ;
}

double f6(double x){
  return exp(-M_PI*x*x);
}

int main(void){
  printf("1, f1, %.15e\n",Qh01(f1,pow(2,-6))-1);
  printf("2, f2, %.15e\n",Qh01(f2,pow(2,-6))*2-1);
  printf("3, f3, %.15e\n",Qh01(f3,pow(2,-6))*100-1);
  printf("4, f4, %.15e\n",Qh01(f4,pow(2,-6))/(exp(1)-1)-1);
  printf("5, f5, %.15e\n",Qhinfinf(f5,pow(2,-6))/M_PI-1);
  printf("6, f6, %.15e\n",Qhinfinf(f6,pow(2,-6))-1);
  printf("7, f5, %.15e\n",Qhainf(f5,0,pow(2,-6))/M_PI*2-1);
  printf("8, f5, %.15e\n",Qhainf(f5,1,pow(2,-6))/M_PI*4-1);
  printf("9, f6, %.15e\n",Qhinfb(f6,0,pow(2,-6))-0.5);
  printf("10, f6, %.15e\n",Qhinfb(f6,1,pow(2,-6))/((1+erf(sqrt(M_PI)))/2)-1);
  return 0;
}

/* 
	Compile in vim on linux mint
	:w | ! gcc qh_1.c -lm && ./a.out

	Sample run.
/usr/bin/ld: warning: /tmp/cctdQsZ5.o: requires executable stack (because the .note.GNU-stack section is executable)
1, f1, -1.110223024625157e-16
2, f2, 2.220446049250313e-16
3, f3, -6.661338147750939e-16
4, f4, 6.661338147750939e-16
5, f5, 6.661338147750939e-16
6, f6, 0.000000000000000e+00
7, f5, 6.661338147750939e-16
8, f5, -6.661338147750939e-16
9, f6, -1.665334536937735e-16
10, f6, 2.220446049250313e-16

*/
